File: library/story_teller.py
#!/usr/bin/env python3
"""
Story Teller: User Interaction Interface

This module provides an interactive interface for users to explore
the system's metaphorical and technical narratives.
"""

from typing import List, Dict, Any
from library.librarian import Librarian

class StoryTeller:
    def __init__(self, librarian: Librarian = None):
        """
        Initialize the StoryTeller with an optional Librarian instance.
        
        Args:
            librarian (Librarian, optional): Librarian instance to use. 
                                             Creates a new one if not provided.
        """
        self.librarian = librarian or Librarian()
    
    def interactive_narrative(self) -> None:
        """
        Provide an interactive narrative exploration of the system.
        """
        print("Welcome to the Nested Librarian System")
        print("Exploring the realms of technological interoperability...\n")
        
        while True:
            print("Choose an exploration path:")
            print("1. System Overview")
            print("2. Technology Gate Details")
            print("3. Creative Principles")
            print("4. Exit")
            
            choice = input("Enter your choice (1-4): ")
            
            if choice == '1':
                print(self.librarian.tell_story())
            elif choice == '2':
                self._explore_technologies()
            elif choice == '3':
                self._explore_creative_principles()
            elif choice == '4':
                print("Farewell, traveler of technological realms!")
                break
            else:
                print("Invalid choice. Please try again.")
    
    def _explore_technologies(self) -> None:
        """
        Explore and display details about system technologies.
        """
        technologies = self.librarian.registry.get('components', {}).get('gate', {}).get('technologies', [])
        
        print("\n--- Technology Gate Details ---")
        for tech in technologies:
            print(f"\n{tech['name']} Technology:")
            print(f"  Role: {tech['role']}")
            print(f"  Version: {tech['version']}")
    
    def _explore_creative_principles(self) -> None:
        """
        Display and explain the system's creative principles.
        """
        principles = self.librarian.registry.get('creative_principles', [])
        
        print("\n--- Creative Principles ---")
        for principle in principles:
            for name, details in principle.items():
                print(f"\n{name.replace('_', ' ').title()}:")
                print(f"  {details['description']}")
    
    def generate_narrative(self, context: str = None) -> str:
        """
        Generate a narrative based on the system's current state.
        
        Args:
            context (str, optional): Specific context for the narrative.
        
        Returns:
            str: Generated narrative.
        """
        return self.librarian.tell_story(context)

def main():
    """
    Main entry point for the StoryTeller.
    """
    story_teller = StoryTeller()
    story_teller.interactive_narrative()

if __name__ == "__main__":
    main() 

File: library/StoryTeller.java
import java.util.List;
import java.util.Map;
import java.util.Scanner;

/**
 * Story Teller: User Interaction Interface
 *
 * This module provides an interactive interface for users to explore
 * the system's metaphorical and technical narratives.
 */
public class StoryTeller {

    private Librarian librarian;

    /**
     * Initialize the StoryTeller with an optional Librarian instance.
     *
     * @param librarian Librarian instance to use. Creates a new one if not provided.
     */
    public StoryTeller(Librarian librarian) {
        this.librarian = (librarian != null) ? librarian : new Librarian();
    }

    /**
     * Default constructor creating a new Librarian.
     */
    public StoryTeller() {
        this(new Librarian());
    }

    /**
     * Provide an interactive narrative exploration of the system.
     */
    public void interactiveNarrative() {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Welcome to the Nested Librarian System");
        System.out.println(
            "Exploring the realms of technological interoperability...\n"
        );

        while (true) {
            System.out.println("Choose an exploration path:");
            System.out.println("1. System Overview");
            System.out.println("2. Technology Gate Details");
            System.out.println("3. Creative Principles");
            System.out.println("4. Exit");

            System.out.print("Enter your choice (1-4): ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.println(librarian.tellStory());
                    break;
                case "2":
                    exploreTechnologies();
                    break;
                case "3":
                    exploreCreativePrinciples();
                    break;
                case "4":
                    System.out.println(
                        "Farewell, traveler of technological realms!"
                    );
                    return;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }

    /**
     * Explore and display details about system technologies.
     */
    private void exploreTechnologies() {
        Map<String, Object> registry = librarian.getRegistry();
        Map<String, Object> components = (Map<
                String,
                Object
            >) registry.getOrDefault("components", Map.of());
        Map<String, Object> gate = (Map<
                String,
                Object
            >) components.getOrDefault("gate", Map.of());
        List<Map<String, Object>> technologies = (List<
                Map<String, Object>
            >) gate.getOrDefault("technologies", List.of());

        System.out.println("\n--- Technology Gate Details ---");
        for (Map<String, Object> tech : technologies) {
            System.out.println(
                "\n" +
                tech.getOrDefault("name", "Unknown Technology") +
                " Technology:"
            );
            System.out.println(
                "  Role: " + tech.getOrDefault("role", "Unspecified")
            );
            System.out.println(
                "  Version: " + tech.getOrDefault("version", "Unknown")
            );
        }
    }

    /**
     * Display and explain the system's creative principles.
     */
    private void exploreCreativePrinciples() {
        Map<String, Object> registry = librarian.getRegistry();
        List<Map<String, Object>> principles = (List<
                Map<String, Object>
            >) registry.getOrDefault("creative_principles", List.of());

        System.out.println("\n--- Creative Principles ---");
        for (Map<String, Object> principle : principles) {
            for (Map.Entry<String, Object> entry : principle.entrySet()) {
                String name = entry.getKey().replace("_", " ");
                Map<String, Object> details = (Map<
                        String,
                        Object
                    >) entry.getValue();

                System.out.println(
                    "\n" +
                    name.substring(0, 1).toUpperCase() +
                    name.substring(1) +
                    ":"
                );
                System.out.println(
                    "  " +
                    details.getOrDefault(
                        "description",
                        "No description available"
                    )
                );
            }
        }
    }

    /**
     * Generate a narrative based on the system's current state.
     *
     * @param context Specific context for the narrative
     * @return Generated narrative
     */
    public String generateNarrative(String context) {
        return librarian.tellStory(context);
    }

    /**
     * Generate a narrative without a specific context.
     *
     * @return Generated narrative
     */
    public String generateNarrative() {
        return librarian.tellStory();
    }

    /**
     * Main entry point for the StoryTeller.
     *
     * @param args Command-line arguments
     */
    public static void main(String[] args) {
        StoryTeller storyTeller = new StoryTeller();
        storyTeller.interactiveNarrative();
    }
}
